tools/distpartitioning/dist_lookup.py

```diff
@@ -0,0 +1,289 @@
+
+        #Split the input global_nids into a list of lists where each list will be 
+        #sent to the respective rank/process
+        #We also need to store the indices, in the indices_list, so that we can re-order
+        #the final result (partition-ids) in the same order as the global-nids (function argument)
+        send_list = []
+        indices_list = []
+        for idx in range(self.world_size):
+            cond = service_owners == idx
+            idxes = np.where(cond)
+            ll = global_nids[idxes[0]]
+            send_list.append(torch.from_numpy(ll))
+            indices_list.append(idxes[0])
+        assert len(np.concatenate(indices_list)) == len(global_nids)
+        assert np.all(np.sort(np.concatenate(indices_list)) == np.arange(len(global_nids)))
+
+        #send the request to everyone else.
+        #As a result of this operation, the current process also receives a list of lists
+        #from all the other processes. 
+        #These lists are global-node-ids whose global-node-ids <-> partition-id mappings 
+        #are owned/stored by the current process
+        owner_req_list = alltoallv_cpu(self.rank, self.world_size, send_list)
+
+        #create the response list here for each of the request list received in the previous
+        #step. Populate the respective partition-ids in this response lists appropriately
+        out_list = []
+        for idx in range(self.world_size):
+            if owner_req_list[idx] is None:
+                out_list.append(torch.empty((0,), dtype=torch.int64))
+                continue
+            #get the node_type_ids and per_type_nids for the incoming global_nids
+            ntype_ids, type_nids = self.id_map(owner_req_list[idx].numpy())
+            nypte_ids, type_nids = ntype_ids.numpy(), type_nids.numpy()
+
+            #lists to store partition-ids for the incoming global-nids
+            type_id_lookups = []
+            local_order_idx = []
+            for tid in range(len(self.partid_list)):
+                cond = ntype_ids == tid
+                local_order_idx.append(np.where(cond)[0])
+                global_type_nids = type_nids[cond]
+                if len(global_type_nids) <= 0:
+                    continue
+
+                local_type_nids = global_type_nids - self.type_nid_begin[tid]
+
+                assert np.all(local_type_nids >= 0)
+                assert np.all(local_type_nids <= (self.type_nid_end[tid] + 1 - self.type_nid_begin[tid]))
+
+                cur_owners = self.partid_list[tid][local_type_nids]
+                type_id_lookups.append(cur_owners)
+
+            if len(type_id_lookups) <= 0:
+                out_list.append(torch.empty((0,), dtype=torch.int64))
+            else:
+                #now reorder results for each request
+                sort_order_idx = np.argsort(np.concatenate(local_order_idx))
+                lookups = np.concatenate(type_id_lookups)[sort_order_idx]
```

Comment:
the code is very difficult to understand. is there anyway to simplify the code?

Reply:
I agree. The code is a little bit duplicated between the functions 'get_partition_ids' and 'get_shuffle_nids' -- only difference is the way the mapping is done with each function call. Will look into ways for providing a function pointer as inputs to these functions so that the common part can be put into a function -- sends and receives messages, does the mapping, sends/receives messages. Once this is done, 'get_partition_ids' and 'get_shuffle_ids' can use that common function and passing a function pointer for mapping purposes. 

If you have any other ideas... please share